satellite/repair/checker/rangedloop.go

```diff
@@ -0,0 +1,411 @@
+	counter4 *monkit.Counter
+	counter5 *monkit.Counter
+}
+
+// NewRangedLoopObserver creates new checker observer instance.
+func NewRangedLoopObserver(logger *zap.Logger, repairQueue queue.RepairQueue, overlay *overlay.Service, config Config) rangedloop.Observer {
+	return &RangedLoopObserver{
+		logger: logger,
+
+		repairQueue:          repairQueue,
+		nodestate:            NewReliabilityCache(overlay, config.ReliabilityCacheStaleness),
+		statsCollector:       newStatsCollector(),
+		repairOverrides:      config.RepairOverrides.GetMap(),
+		nodeFailureRate:      config.NodeFailureRate,
+		repairQueueBatchSize: config.RepairQueueInsertBatchSize,
+
+		TotalStats: aggregateStats{},
+
+		counter1: mon.Counter("rl_remote_segments_over_threshold_1"),
+		counter2: mon.Counter("rl_remote_segments_over_threshold_2"),
+		counter3: mon.Counter("rl_remote_segments_over_threshold_3"),
+		counter4: mon.Counter("rl_remote_segments_over_threshold_4"),
+		counter5: mon.Counter("rl_remote_segments_over_threshold_5"),
+	}
+}
+
+// getNodesEstimate updates the estimate of the total number of nodes. It is guaranteed
+// to return a number greater than 0 when the error is nil.
+//
+// We can't calculate this upon first starting a Ranged Loop Observer, because there may not be any
+// nodes yet. We expect that there will be nodes before there are segments, though.
+func (observer *RangedLoopObserver) getNodesEstimate(ctx context.Context) (int, error) {
+	// this should be safe to call frequently; it is an efficient caching lookup.
+	totalNumNodes, err := observer.nodestate.NumNodes(ctx)
+	if err != nil {
+		// We could proceed here by returning the last good value, or by returning a fallback
+		// constant estimate, like "20000", and we'd probably be fine, but it would be better
+		// not to have that happen silently for too long. Also, if we can't get this from the
+		// database, we probably can't modify the injured segments queue, so it won't help to
+		// proceed with this repair operation.
+		return 0, err
+	}
+
+	if totalNumNodes == 0 {
+		return 0, Error.New("segment health is meaningless: there are no nodes")
+	}
+
+	return totalNumNodes, nil
+}
+
+func (observer *RangedLoopObserver) createInsertBuffer() *queue.InsertBuffer {
+	return queue.NewInsertBuffer(observer.repairQueue, observer.repairQueueBatchSize)
+}
+
+// CompareInjuredSegment compares stream id of injured segment, for testing purposes only.
+func (observer *RangedLoopObserver) CompareInjuredSegment(ctx context.Context, streamID uuid.UUID) error {
+	injuredSegment, err := observer.repairQueue.Select(ctx)
+	if err != nil {
+		return err
+	}
+
+	if injuredSegment.StreamID != streamID {
+		return errs.New("streamID of injured segment does not match")
+	}
+
+	return nil
+}
+
+// CompareStats compares total stats with given data, used only for tests.
+func (observer *RangedLoopObserver) CompareStats(objChecked int64, remoteSegmentsChecked int64, remoteSegmentsNeedingRepair int64,
```

Comment:
If thats only for tests maybe we could move it to test? TotalStats should accessible there

Reply:
what with this comment?

Reply:
aggregated stats are private struct with all private fields, I need to change it to public to be able to move this method to test package, should I?

Reply:
ah right, then lets add `Compare()` to aggregateStats type. Then it should be possible do something like `observer.TotalStats.Compare(45, 45, 5, 5, 0, 0, nil))`